doc/config_yang_validation/config_db_yang_validation.md

```diff
@@ -0,0 +1,178 @@
+    - [4.2 YANG Validation Implementation](#yang-validation-implementation)
+    - [4.3 Additional Features](#additional-features)
+
+
+### Revision
+
+|  Rev  |  Date   |      Author      | Change Description |
+| :---: | :-----: | :--------------: | ------------------ |
+|  1.0  | 08/2022 |     Isabel Li    | Phase 1 Design     |
+
+## Project Overview  
+
+### Problem Statement
+Currently, SONiC config fields updates are validated ad-hoc, and much of the validation criteria are duplicated both in python (hard-coded validation specs) and in YANG models
+
+### Goal
+Refactor code to utilize preexisting YANG models for config validation, instead of duplicating the validation specs in both YANG and in python
+
+### Scope
+
+#### Python
+1.	SONiC config fields updated via config CLI (e.g. config portchannel add <…>). A comprehensive list of config fields in this project’s scope is as follows, defined in sonic-utilities/config directory:
+    - 	AAA 
+    - 	ACL 
+    - 	BGP
+    - 	Console 
+    - 	DHCP Relay 
+    - 	Drop Counter 
+    - 	Dynamic Buffer Management
+    - 	ECN
+    - 	Feature
+    - 	Interface 
+    - 	Interface Naming Mode
+    - 	Interface Vrf Binding
+    - 	IPv6 Link Local
+    - 	Kubernetes
+    - 	Linux Kernel Dump
+    - 	Loopback Interfaces
+    - 	VRF
+    - 	Management VRF
+    - 	Mirroring
+    - 	Muxcable
+    - 	NAT Enable/Disable YANG Validation
+    - 	NTP
+    - 	NVGRE
+    - 	PBH
+    - 	Platform Component Firmware
+    - 	PortChannel
+    - 	QoS
+    - 	sFlow
+    - 	SNMP
+    - 	Subinterfaces
+    - 	Syslog
+    - 	VLAN
+    - 	Warm Restart
+    - 	Watermark
+    - 	ZTP
+3.	ACL rules: incremental update, delete
+4.	pfc
+5.	crm
+6.	mclag
+7.	counterpoll 
+8.	Scripts (python)
+a.	configlet
+b.	db_migrator
+c.	dropconfig
+d.	mellanox_buffer_migrator
+e.	neighbor_advertisor
+f.	null_route_helper
+g.	hostcfgd
+
+#### C++
+1. buffermgrd
+
+
+## Case Study
+
+### Portchannel
+
+PortChannel config CLI is defined in sonic-utilities/config/main.py. In this file, we find the following config validation:  
+![](../../images/config_yang_validation/adhoc_portchannel.png)
+
+However, these specs for validation are also defined in the sonic-portchannel YANG model:  
+![](../../images/config_yang_validation/yang_portchannel.png)
+
+For PortChannel and many of the config fields listed above, the field validation is unnecessarily duplicated in sonic-utilities/config and in YANG models. The goal of this project is to utilize preexisting YANG models to validate config CLI field updates, without unnecessarily defining separate, redundant validation specs ad-hoc in sonic-utilities. 
+
+During the migration process, we will first leave the ad-hoc validation code in-place, and leave an option to configure type of validation used: ad-hoc or YANG validation. Once YANG validation has stabilized and YANG validation coverage has widened, we will remove the preexisting ad-hoc validation code for non performance-sensitive scenarios. 
```

Comment:
Some adhoc validations cannot be accomplished by yang models for e.g. there are no validations in yang to ensure that port added to portchannel is not part of a Vlan. So some adhoc validations need to retained for which yang support is unavailable even though we enable yang validation. Can you please capture this?

Reply:
By "cannot be accomplished by yang models," do you mean that the constraints are currently missing from yang models, or it's not possible to reflect the constraints in yang models? 

I noticed that some adhoc validations are indeed missing from yang models, and I added # TODO comment in code to add the missing constraint(s) to corresponding yang models. I will also create issues to track missing yang model constraints.

Reply:
@isabelmsft Sounds good. For the missing validations which can be added in yang models we can track as issues. However if yang infrastructure cannot support certain types of checks for some reason we need to have adhoc validations for such cases and an explicit comment in the CLI would be better.

Reply:
I see, I added a note describing the limitations of some YANG models and our intent to fill in the gaps where possible.